finat/finiteelementbase.py

```diff
@@ -144,6 +145,110 @@ def point_evaluation(self, order, refcoords, entity=None):
         :param entity: the cell entity on which to tabulate.
         '''
 
+    @abstractproperty
+    def dual_basis(self):
+        '''Returns a tuple where each element of the tuple represents one
+        functional in the dual space. Each functional is represented by
+        a tuple of tuples containing the points (PointSet), a weight tensor
+        which holds the weights for each component of value_shape of functions,
+        and a alpha tensor for extracting the alpha components from ReferenceGrad,
+        sorted by total derivative order.
+
+        For example, a dual basis containing 2 functionals with maximum derivative
+        order of 1 would be represented by:
+
+        .. highlight:: python
+        .. code-block:: python
+
+            ((((point_set_10, weight_tensor_10, alpha_tensor_10),
+               (point_set_11, weight_tensor_11, alpha_tensor_11)), None)
+             (((point_set_20, weight_tensor_20, alpha_tensor_20),
+              ()), None)
+
+        where one of the innermost tuples is empty because there are no evaluations
+        at that order. The `None` is either a placeholder for \'normal\' elements,
+        or a tuple containing the index of expression to extract using gem.Delta
+        (currently just used by TensorFiniteElement).
+        '''
+        # TODO: Add label for type of evaluation?
+
+    def dual_evaluation(self, fn, entity=None):
```

Comment:
Note: In order to get `PhysicallyMappedElement` (aka zany elements) correct, you need to overload `dual_evaluation`.  Morally, one should just get the list of results (functionals all applied to `fn`) and smack that with the transpose of the `PhysicallyMappedElement.basis_transformation` method.  To see how this is done with basis rather than dual evaluation: https://github.com/FInAT/FInAT/blob/2ef800c10efc37236a0e14f672caf9355b88a5d0/finat/physically_mapped.py#L158